File: backend/internal/model/medicamento.go
package model

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// Medicamento representa um registro de uso de medicamento em um lote.
type Medicamento struct {
	ID                  uuid.UUID  `gorm:"type:uuid;primaryKey;default:gen_random_uuid()" json:"id"`
	LoteID              uuid.UUID  `gorm:"type:uuid;not null;index:idx_medicamentos_lote" json:"lote_id" validate:"required"`
	DataInicio          time.Time  `gorm:"type:date;not null" json:"data_inicio" validate:"required"`
	DataFim             *time.Time `gorm:"type:date" json:"data_fim,omitempty"`
	Medicamento         string     `gorm:"type:varchar(200);not null;column:medicamento" json:"medicamento" validate:"required,max=200"`
	Dosagem             *string    `gorm:"type:varchar(100)" json:"dosagem,omitempty" validate:"omitempty,max=100"`
	Via                 *string    `gorm:"type:varchar(50)" json:"via,omitempty" validate:"omitempty,max=50"`
	PeriodoCarenciaDias *int       `gorm:"type:integer;default:0" json:"periodo_carencia_dias,omitempty" validate:"omitempty,gte=0"`
	Responsavel         *string    `gorm:"type:varchar(200)" json:"responsavel,omitempty" validate:"omitempty,max=200"`
	Observacao          *string    `gorm:"type:text" json:"observacao,omitempty"`
	CreatedAt           time.Time  `gorm:"not null;default:now()" json:"created_at"`

	// Relacionamentos
	Lote Lote `gorm:"foreignKey:LoteID" json:"-"`
}

func (Medicamento) TableName() string {
	return "medicamentos"
}

func (m *Medicamento) BeforeCreate(tx *gorm.DB) error {
	if m.ID == uuid.Nil {
		m.ID = uuid.New()
	}
	return nil
}

File: flutter_app/lib/features/sanidade/domain/entities/medicamento.dart
/// Registro de medicamento aplicado em um lote.
///
/// Matches backend DTO MedicamentoResponse:
/// - medicamento (not nome)
/// - via (not via_administracao)
/// - data_inicio, data_fim
/// - periodo_carencia_dias
/// - em_carencia
class Medicamento {
  final String id;
  final String loteId;
  final String dataInicio;
  final String? dataFim;
  final String medicamento;
  final String? dosagem;
  final String? via;
  final int? periodoCarenciaDias;
  final String? responsavel;
  final String? observacao;
  final bool emCarencia;
  final String createdAt;

  const Medicamento({
    required this.id,
    required this.loteId,
    required this.dataInicio,
    this.dataFim,
    required this.medicamento,
    this.dosagem,
    this.via,
    this.periodoCarenciaDias,
    this.responsavel,
    this.observacao,
    this.emCarencia = false,
    required this.createdAt,
  });

  factory Medicamento.fromJson(Map<String, dynamic> json) {
    return Medicamento(
      id: json['id'] as String,
      loteId: json['lote_id'] as String,
      dataInicio: json['data_inicio'] as String,
      dataFim: json['data_fim'] as String?,
      medicamento: json['medicamento'] as String,
      dosagem: json['dosagem'] as String?,
      via: json['via'] as String?,
      periodoCarenciaDias: (json['periodo_carencia_dias'] as num?)?.toInt(),
      responsavel: json['responsavel'] as String?,
      observacao: json['observacao'] as String?,
      emCarencia: json['em_carencia'] as bool? ?? false,
      createdAt: json['created_at'] as String,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'lote_id': loteId,
      'data_inicio': dataInicio,
      if (dataFim != null) 'data_fim': dataFim,
      'medicamento': medicamento,
      if (dosagem != null) 'dosagem': dosagem,
      if (via != null) 'via': via,
      if (periodoCarenciaDias != null)
        'periodo_carencia_dias': periodoCarenciaDias,
      if (responsavel != null) 'responsavel': responsavel,
      if (observacao != null) 'observacao': observacao,
      'em_carencia': emCarencia,
      'created_at': createdAt,
    };
  }
}
